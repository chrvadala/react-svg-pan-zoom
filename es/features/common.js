var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

import { TOOL_NONE, MODE_IDLE } from '../constants';
import { identity, fromObject, inverse, applyToPoint, transform, translate, scale } from 'transformation-matrix';

/**
 * Obtain default value
 * @returns {Object}
 */
export function getDefaultValue(viewerWidth, viewerHeight, SVGWidth, SVGHeight) {
  return set({}, _extends({}, identity(), {
    version: 2,
    mode: MODE_IDLE,
    focus: false,
    pinchPointDistance: null,
    prePinchMode: null,
    viewerWidth: viewerWidth,
    viewerHeight: viewerHeight,
    SVGWidth: SVGWidth,
    SVGHeight: SVGHeight,
    startX: null,
    startY: null,
    endX: null,
    endY: null,
    miniatureOpen: true
  }));
}

/**
 * Change value
 * @param value
 * @param change
 * @returns {Object}
 */
export function set(value, change) {
  value = Object.assign({}, value, change);
  return Object.freeze(value);
}

/**
 * value valid check
 * @param value
 */
export function isValueValid(value) {
  return value !== null && (typeof value === 'undefined' ? 'undefined' : _typeof(value)) === 'object' && value.hasOwnProperty('version');
}

/**
 * Export x,y coords relative to SVG
 * @param value
 * @param viewerX
 * @param viewerY
 * @returns {*|{x, y}|{x: number, y: number}}
 */
export function getSVGPoint(value, viewerX, viewerY) {
  var matrix = fromObject(value);

  var inverseMatrix = inverse(matrix);
  return applyToPoint(inverseMatrix, { x: viewerX, y: viewerY });
}

/**
 * Decompose matrix from value
 * @param value
 * @returns {{scaleFactor: number, translationX: number, translationY: number}}
 */
export function decompose(value) {
  var matrix = fromObject(value);

  return {
    scaleFactor: matrix.a,
    translationX: matrix.e,
    translationY: matrix.f
  };
}

/**
 *
 * @param value
 * @param focus
 * @returns {Object}
 */
export function setFocus(value, focus) {
  return set(value, { focus: focus });
}

/**
 *
 * @param value
 * @param viewerWidth
 * @param viewerHeight
 * @returns {Object}
 */
export function setViewerSize(value, viewerWidth, viewerHeight) {
  return set(value, { viewerWidth: viewerWidth, viewerHeight: viewerHeight });
}

/**
 *
 * @param value
 * @param SVGWidth
 * @param SVGHeight
 * @returns {Object}
 */
export function setSVGSize(value, SVGWidth, SVGHeight) {
  return set(value, { SVGWidth: SVGWidth, SVGHeight: SVGHeight });
}

/**
 *
 * @param value
 * @param SVGPointX
 * @param SVGPointY
 * @param zoomLevel
 * @returns {Object}
 */
export function setPointOnViewerCenter(value, SVGPointX, SVGPointY, zoomLevel) {
  var viewerWidth = value.viewerWidth,
      viewerHeight = value.viewerHeight;


  var matrix = transform(translate(-SVGPointX + viewerWidth / 2, -SVGPointY + viewerHeight / 2), //4
  translate(SVGPointX, SVGPointY), //3
  scale(zoomLevel, zoomLevel), //2
  translate(-SVGPointX, -SVGPointY) //1
  );

  return set(value, _extends({
    mode: MODE_IDLE
  }, matrix));
}

/**
 *
 * @param value
 * @returns {Object}
 */
export function reset(value) {
  return set(value, _extends({
    mode: MODE_IDLE
  }, identity()));
}

/**
 *
 * @param value
 * @returns {Object}
 */
export function resetMode(value) {
  return set(value, {
    mode: MODE_IDLE,
    startX: null,
    startY: null,
    endX: null,
    endY: null
  });
}