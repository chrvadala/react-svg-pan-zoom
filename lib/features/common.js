'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.getDefaultValue = getDefaultValue;
exports.set = set;
exports.isValueValid = isValueValid;
exports.getSVGPoint = getSVGPoint;
exports.decompose = decompose;
exports.setFocus = setFocus;
exports.setViewerSize = setViewerSize;
exports.setSVGSize = setSVGSize;
exports.setPointOnViewerCenter = setPointOnViewerCenter;
exports.reset = reset;
exports.resetMode = resetMode;

var _constants = require('../constants');

var _transformationMatrix = require('transformation-matrix');

/**
 * Obtain default value
 * @returns {Object}
 */
function getDefaultValue(viewerWidth, viewerHeight, SVGWidth, SVGHeight) {
  return set({}, _extends({}, (0, _transformationMatrix.identity)(), {
    version: 2,
    mode: _constants.MODE_IDLE,
    focus: false,
    pinchPointDistance: null,
    prePinchMode: null,
    viewerWidth: viewerWidth,
    viewerHeight: viewerHeight,
    SVGWidth: SVGWidth,
    SVGHeight: SVGHeight,
    startX: null,
    startY: null,
    endX: null,
    endY: null,
    miniatureOpen: true
  }));
}

/**
 * Change value
 * @param value
 * @param change
 * @returns {Object}
 */
function set(value, change) {
  value = Object.assign({}, value, change);
  return Object.freeze(value);
}

/**
 * value valid check
 * @param value
 */
function isValueValid(value) {
  return value !== null && (typeof value === 'undefined' ? 'undefined' : _typeof(value)) === 'object' && value.hasOwnProperty('version');
}

/**
 * Export x,y coords relative to SVG
 * @param value
 * @param viewerX
 * @param viewerY
 * @returns {*|{x, y}|{x: number, y: number}}
 */
function getSVGPoint(value, viewerX, viewerY) {
  var matrix = (0, _transformationMatrix.fromObject)(value);

  var inverseMatrix = (0, _transformationMatrix.inverse)(matrix);
  return (0, _transformationMatrix.applyToPoint)(inverseMatrix, { x: viewerX, y: viewerY });
}

/**
 * Decompose matrix from value
 * @param value
 * @returns {{scaleFactor: number, translationX: number, translationY: number}}
 */
function decompose(value) {
  var matrix = (0, _transformationMatrix.fromObject)(value);

  return {
    scaleFactor: matrix.a,
    translationX: matrix.e,
    translationY: matrix.f
  };
}

/**
 *
 * @param value
 * @param focus
 * @returns {Object}
 */
function setFocus(value, focus) {
  return set(value, { focus: focus });
}

/**
 *
 * @param value
 * @param viewerWidth
 * @param viewerHeight
 * @returns {Object}
 */
function setViewerSize(value, viewerWidth, viewerHeight) {
  return set(value, { viewerWidth: viewerWidth, viewerHeight: viewerHeight });
}

/**
 *
 * @param value
 * @param SVGWidth
 * @param SVGHeight
 * @returns {Object}
 */
function setSVGSize(value, SVGWidth, SVGHeight) {
  return set(value, { SVGWidth: SVGWidth, SVGHeight: SVGHeight });
}

/**
 *
 * @param value
 * @param SVGPointX
 * @param SVGPointY
 * @param zoomLevel
 * @returns {Object}
 */
function setPointOnViewerCenter(value, SVGPointX, SVGPointY, zoomLevel) {
  var viewerWidth = value.viewerWidth,
      viewerHeight = value.viewerHeight;


  var matrix = (0, _transformationMatrix.transform)((0, _transformationMatrix.translate)(-SVGPointX + viewerWidth / 2, -SVGPointY + viewerHeight / 2), //4
  (0, _transformationMatrix.translate)(SVGPointX, SVGPointY), //3
  (0, _transformationMatrix.scale)(zoomLevel, zoomLevel), //2
  (0, _transformationMatrix.translate)(-SVGPointX, -SVGPointY) //1
  );

  return set(value, _extends({
    mode: _constants.MODE_IDLE
  }, matrix));
}

/**
 *
 * @param value
 * @returns {Object}
 */
function reset(value) {
  return set(value, _extends({
    mode: _constants.MODE_IDLE
  }, (0, _transformationMatrix.identity)()));
}

/**
 *
 * @param value
 * @returns {Object}
 */
function resetMode(value) {
  return set(value, {
    mode: _constants.MODE_IDLE,
    startX: null,
    startY: null,
    endX: null,
    endY: null
  });
}