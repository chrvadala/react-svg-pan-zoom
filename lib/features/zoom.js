'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.zoom = zoom;
exports.fitSelection = fitSelection;
exports.fitToViewer = fitToViewer;
exports.zoomOnViewerCenter = zoomOnViewerCenter;
exports.startZooming = startZooming;
exports.updateZooming = updateZooming;
exports.stopZooming = stopZooming;

var _transformationMatrix = require('transformation-matrix');

var _constants = require('../constants');

var _common = require('./common');

var _utils = require('../utils');

function zoom(value, SVGPointX, SVGPointY, scaleFactor, extraData) {

  var matrix = (0, _transformationMatrix.transform)((0, _transformationMatrix.fromObject)(value), (0, _transformationMatrix.translate)(SVGPointX, SVGPointY), (0, _transformationMatrix.scale)(scaleFactor, scaleFactor), (0, _transformationMatrix.translate)(-SVGPointX, -SVGPointY));

  return (0, _common.set)(value, (0, _common.set)(_extends({
    mode: _constants.MODE_IDLE
  }, matrix, {
    startX: null,
    startY: null,
    endX: null,
    endY: null
  }), extraData));
}

function fitSelection(value, selectionSVGPointX, selectionSVGPointY, selectionWidth, selectionHeight) {
  var viewerWidth = value.viewerWidth,
      viewerHeight = value.viewerHeight;


  var scaleX = viewerWidth / selectionWidth;
  var scaleY = viewerHeight / selectionHeight;

  var scaleLevel = Math.min(scaleX, scaleY);

  var matrix = (0, _transformationMatrix.transform)((0, _transformationMatrix.scale)(scaleLevel, scaleLevel), //2
  (0, _transformationMatrix.translate)(-selectionSVGPointX, -selectionSVGPointY) //1
  );

  return (0, _common.set)(value, _extends({
    mode: _constants.MODE_IDLE
  }, matrix, {
    startX: null,
    startY: null,
    endX: null,
    endY: null
  }));
}

function fitToViewer(value) {
  return fitSelection(value, 0, 0, value.SVGWidth, value.SVGHeight);
}

function zoomOnViewerCenter(value, scaleFactor) {
  var viewerWidth = value.viewerWidth,
      viewerHeight = value.viewerHeight;

  var SVGPoint = (0, _common.getSVGPoint)(value, viewerWidth / 2, viewerHeight / 2);
  return zoom(value, SVGPoint.x, SVGPoint.y, scaleFactor);
}

function startZooming(value, viewerX, viewerY) {
  return (0, _common.set)(value, {
    mode: _constants.MODE_ZOOMING,
    startX: viewerX,
    startY: viewerY,
    endX: viewerX,
    endY: viewerY
  });
}

function updateZooming(value, viewerX, viewerY) {
  if (value.mode !== _constants.MODE_ZOOMING) throw new Error('update selection not allowed in this mode ' + value.mode);

  return (0, _common.set)(value, {
    endX: viewerX,
    endY: viewerY
  });
}

function stopZooming(value, viewerX, viewerY, scaleFactor) {
  var startX = value.startX,
      startY = value.startY,
      endX = value.endX,
      endY = value.endY;


  var start = (0, _common.getSVGPoint)(value, startX, startY);
  var end = (0, _common.getSVGPoint)(value, endX, endY);

  if (Math.abs(startX - endX) > 7 && Math.abs(startY - endY) > 7) {
    var box = (0, _utils.calculateBox)(start, end);
    return fitSelection(value, box.x, box.y, box.width, box.height);
  } else {
    var SVGPoint = (0, _common.getSVGPoint)(value, viewerX, viewerY);
    return zoom(value, SVGPoint.x, SVGPoint.y, scaleFactor);
  }
}