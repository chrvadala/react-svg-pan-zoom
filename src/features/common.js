import {TOOL_NONE, MODE_IDLE} from '../constants';
import {
  identity,
  fromObject,
  inverse,
  applyToPoint,
  transform,
  translate,
  scale
} from 'transformation-matrix';

/**
 * Obtain default value
 * @returns {Object}
 */
export function getDefaultValue(viewerWidth, viewerHeight, SVGWidth, SVGHeight) {
  return set({}, {
    ...identity(),
    version: 2,
    mode: MODE_IDLE,
    focus: false,
    pinchPointDistance: null,
    prePinchMode: null,
    viewerWidth,
    viewerHeight,
    SVGWidth,
    SVGHeight,
    startX: null,
    startY: null,
    endX: null,
    endY: null,
    miniatureOpen: true
  });
}

/**
 * Change value
 * @param value
 * @param change
 * @returns {Object}
 */
export function set(value, change) {
  value = Object.assign({}, value, change);
  return Object.freeze(value);
}

/**
 * value valid check
 * @param value
 */
export function isValueValid(value) {
  return value !== null
    && typeof value === 'object'
    && value.hasOwnProperty('version');
}

/**
 * Export x,y coords relative to SVG
 * @param value
 * @param viewerX
 * @param viewerY
 * @returns {*|{x, y}|{x: number, y: number}}
 */
export function getSVGPoint(value, viewerX, viewerY) {
  let matrix = fromObject(value);

  let inverseMatrix = inverse(matrix);
  return applyToPoint(inverseMatrix, {x: viewerX, y: viewerY});
}

/**
 * Decompose matrix from value
 * @param value
 * @returns {{scaleFactor: number, translationX: number, translationY: number}}
 */
export function decompose(value) {
  let matrix = fromObject(value);

  return {
    scaleFactor: matrix.a,
    translationX: matrix.e,
    translationY: matrix.f
  }
}

/**
 *
 * @param value
 * @param focus
 * @returns {Object}
 */
export function setFocus(value, focus) {
  return set(value, {focus});
}


/**
 *
 * @param value
 * @param viewerWidth
 * @param viewerHeight
 * @returns {Object}
 */
export function setViewerSize(value, viewerWidth, viewerHeight) {
  return set(value, {viewerWidth, viewerHeight});
}

/**
 *
 * @param value
 * @param SVGWidth
 * @param SVGHeight
 * @returns {Object}
 */
export function setSVGSize(value, SVGWidth, SVGHeight) {
  return set(value, {SVGWidth, SVGHeight});
}

/**
 *
 * @param value
 * @param SVGPointX
 * @param SVGPointY
 * @param zoomLevel
 * @returns {Object}
 */
export function setPointOnViewerCenter(value, SVGPointX, SVGPointY, zoomLevel) {
  let {viewerWidth, viewerHeight} = value;

  let matrix = transform(
    translate(-SVGPointX + viewerWidth / 2, -SVGPointY + viewerHeight / 2),   //4
    translate(SVGPointX, SVGPointY),                                          //3
    scale(zoomLevel, zoomLevel),                                              //2
    translate(-SVGPointX, -SVGPointY)                                         //1
  );

  return set(value, {
    mode: MODE_IDLE,
    ...matrix,
  });
}

/**
 *
 * @param value
 * @returns {Object}
 */
export function reset(value) {
  return set(value, {
    mode: MODE_IDLE,
    ...identity()
  });
}

/**
 *
 * @param value
 * @returns {Object}
 */
export function resetMode(value) {
  return set(value, {
    mode: MODE_IDLE,
    startX: null,
    startY: null,
    endX: null,
    endY: null
  })
}
