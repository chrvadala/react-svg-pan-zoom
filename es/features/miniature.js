import { set } from './common';

export function openMiniature(value) {
  return set(value, {
    miniatureOpen: true
  });
}

export function closeMiniature(value) {
  return set(value, {
    miniatureOpen: false
  });
}