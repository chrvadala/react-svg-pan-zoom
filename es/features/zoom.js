var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

import { transform, fromObject, translate, scale } from 'transformation-matrix';

import { MODE_IDLE, MODE_ZOOMING } from '../constants';
import { set, getSVGPoint } from './common';
import { calculateBox } from '../utils';

export function zoom(value, SVGPointX, SVGPointY, scaleFactor, extraData) {

  var matrix = transform(fromObject(value), translate(SVGPointX, SVGPointY), scale(scaleFactor, scaleFactor), translate(-SVGPointX, -SVGPointY));

  return set(value, set(_extends({
    mode: MODE_IDLE
  }, matrix, {
    startX: null,
    startY: null,
    endX: null,
    endY: null
  }), extraData));
}

export function fitSelection(value, selectionSVGPointX, selectionSVGPointY, selectionWidth, selectionHeight) {
  var viewerWidth = value.viewerWidth,
      viewerHeight = value.viewerHeight;


  var scaleX = viewerWidth / selectionWidth;
  var scaleY = viewerHeight / selectionHeight;

  var scaleLevel = Math.min(scaleX, scaleY);

  var matrix = transform(scale(scaleLevel, scaleLevel), //2
  translate(-selectionSVGPointX, -selectionSVGPointY) //1
  );

  return set(value, _extends({
    mode: MODE_IDLE
  }, matrix, {
    startX: null,
    startY: null,
    endX: null,
    endY: null
  }));
}

export function fitToViewer(value) {
  return fitSelection(value, 0, 0, value.SVGWidth, value.SVGHeight);
}

export function zoomOnViewerCenter(value, scaleFactor) {
  var viewerWidth = value.viewerWidth,
      viewerHeight = value.viewerHeight;

  var SVGPoint = getSVGPoint(value, viewerWidth / 2, viewerHeight / 2);
  return zoom(value, SVGPoint.x, SVGPoint.y, scaleFactor);
}

export function startZooming(value, viewerX, viewerY) {
  return set(value, {
    mode: MODE_ZOOMING,
    startX: viewerX,
    startY: viewerY,
    endX: viewerX,
    endY: viewerY
  });
}

export function updateZooming(value, viewerX, viewerY) {
  if (value.mode !== MODE_ZOOMING) throw new Error('update selection not allowed in this mode ' + value.mode);

  return set(value, {
    endX: viewerX,
    endY: viewerY
  });
}

export function stopZooming(value, viewerX, viewerY, scaleFactor) {
  var startX = value.startX,
      startY = value.startY,
      endX = value.endX,
      endY = value.endY;


  var start = getSVGPoint(value, startX, startY);
  var end = getSVGPoint(value, endX, endY);

  if (Math.abs(startX - endX) > 7 && Math.abs(startY - endY) > 7) {
    var box = calculateBox(start, end);
    return fitSelection(value, box.x, box.y, box.width, box.height);
  } else {
    var SVGPoint = getSVGPoint(value, viewerX, viewerY);
    return zoom(value, SVGPoint.x, SVGPoint.y, scaleFactor);
  }
}