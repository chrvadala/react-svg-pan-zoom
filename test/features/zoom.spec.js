import {getDefaultValue} from "../../src/features/common";
import {
  ALIGN_BOTTOM, ALIGN_CENTER,
  ALIGN_LEFT,
  ALIGN_RIGHT,
  ALIGN_TOP,
  fitToViewer,
  MODE_IDLE,
  MODE_ZOOMING,
  zoom,
  zoomOnViewerCenter
} from "../../src";
import {testBBox} from "../test-utils";
import {startZooming, stopZooming, updateZooming} from "../../src/features/zoom";

// declaring this because Number.EPSILON doesn't work with comparable numbers
const EPSILON = 0.00000001

describe("zoom", () => {
  test("test zoom in", () => {
    const value = getDefaultValue(
      200, 200,       //viewer 200x200
      0, 0, 400, 400, //svg 400x400
    )

    const value1 = zoom(value, 0, 0, 2)
    expect(testBBox(value1)).toEqual([0, 0, 800, 800])
    expect(value1).toMatchSnapshot()

    const value2 = zoom(value1, 0, 0, 2)
    expect(testBBox(value2)).toEqual([0, 0, 1600, 1600])
    expect(value2).toMatchSnapshot()

    const value3 = zoom(value, 100, 100, 2)
    expect(testBBox(value3)).toEqual([-100, -100, 700, 700])

    const value4 = zoom(value, 200, 200, 2)
    expect(testBBox(value4)).toEqual([-200, -200, 600, 600])
  })

  test("test zoom out", () => {
    const value = getDefaultValue(
      200, 200,       //viewer 200x200
      0, 0, 400, 400, //svg 400x400
    )

    const value1 = zoom(value, 0, 0, 1 / 2)
    expect(testBBox(value1)).toEqual([0, 0, 200, 200])
    expect(value1).toMatchSnapshot()

    const value2 = zoom(value1, 0, 0, 1 / 2)
    expect(testBBox(value2)).toEqual([0, 0, 100, 100])
    expect(value2).toMatchSnapshot()

    const value3 = zoom(value, 100, 100, 1 / 2)
    expect(testBBox(value3)).toEqual([50, 50, 250, 250])

    const value4 = zoom(value, 200, 200, 1 / 2)
    expect(testBBox(value4)).toEqual([100, 100, 300, 300])
  })

  test("test min bound", () => {
    const value = getDefaultValue(
      200, 200,       //viewer 200x200
      0, 0, 400, 400, //svg 400x400
      1 / 2, 4          //min/max scaleFactor 1/2 - 4
    )

    const value1 = zoom(value, 0, 0, 4 + EPSILON)
    expect(value1).toBe(value) //should be unchanged

    const value2 = zoom(value, 0, 0, 4)
    expect(testBBox(value2)).toEqual([0, 0, 1600, 1600])
  })

  test("test min bound", () => {
    const value = getDefaultValue(
      200, 200,       //viewer 200x200
      0, 0, 400, 400, //svg 400x400
      1 / 2, 4          //min/max scaleFactor 1/2 - 4
    )

    const value1 = zoom(value, 0, 0, 1 / 2 - EPSILON)
    expect(value1).toBe(value) //should be unchanged

    const value2 = zoom(value, 0, 0, 1 / 2)
    expect(testBBox(value2)).toEqual([0, 0, 200, 200])
  })
})

describe("fitSelection", () => {
  //TODO
})

describe("fitToViewer", () => {
  test("rect with w>h", () => {
    const value = getDefaultValue(
      200, 100,       //viewer 200x100
      0, 0, 50, 50, //svg 50x50
    )

    const value1 = fitToViewer(value, ALIGN_LEFT, ALIGN_TOP)
    expect(testBBox(value1)).toEqual([0, 0, 100, 100])

    const value2 = fitToViewer(value, ALIGN_CENTER, ALIGN_TOP)
    expect(testBBox(value2)).toEqual([50, 0, 150, 100])

    const value3 = fitToViewer(value, ALIGN_RIGHT, ALIGN_TOP)
    expect(testBBox(value3)).toEqual([100, 0, 200, 100])
  })

  test("rect with w<h", () => {
    const value = getDefaultValue(
      100, 200,       //viewer 100x200
      0, 0, 50, 50, //svg 50x50
    )

    const value1 = fitToViewer(value, ALIGN_LEFT, ALIGN_TOP)
    expect(testBBox(value1)).toEqual([0, 0, 100, 100])

    const value2 = fitToViewer(value, ALIGN_LEFT, ALIGN_CENTER)
    expect(testBBox(value2)).toEqual([0, 50, 100, 150])

    const value3 = fitToViewer(value, ALIGN_LEFT, ALIGN_BOTTOM)
    expect(testBBox(value3)).toEqual([0, 100, 100, 200])
  })
})

test("zoomOnViewerCenter", () => {
  const value = getDefaultValue(
    200, 200,       //viewer 200x200
    0, 0, 200, 200, //svg 200x200
  )

  const value1 = zoomOnViewerCenter(value, 2)
  expect(testBBox(value1)).toEqual([-100, -100, 300, 300])

  const value2 = zoomOnViewerCenter(value, 1 / 2)
  expect(testBBox(value2)).toEqual([50, 50, 150, 150])
})

test("zoom lifecycle", () => {
  const value = getDefaultValue(
    200, 200,       //viewer 200x200
    0, 0, 200, 200, //svg 200x200
  )

  const value1 = startZooming(value, 0, 0)
  expect(value1).toMatchObject({mode: MODE_ZOOMING, startX: 0, startY: 0})

  const value2 = updateZooming(value1, 100, 100)
  expect(value2).toMatchObject({endX: 100, endY: 100})

  //zoom on point
  const value3 = stopZooming(value2, 0, 0, 2)
  expect(testBBox(value3)).toEqual([0, 0, 400, 400])
  expect(value3).toMatchObject({mode: MODE_IDLE})
  expect(value3).toMatchSnapshot()

  //zoom on area
  const value4 = stopZooming(value2, 50, 50)
  expect(testBBox(value4)).toEqual([0, 0, 800, 800])
  expect(value4).toMatchObject({mode: MODE_IDLE})
  expect(value4).toMatchSnapshot()
})
