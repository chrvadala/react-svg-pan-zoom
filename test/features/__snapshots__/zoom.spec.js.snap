// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zoom lifecycle 1`] = `
Object {
  "SVGHeight": 200,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 200,
  "a": 2,
  "b": 0,
  "c": 0,
  "d": 2,
  "e": 0,
  "endX": null,
  "endY": null,
  "f": 0,
  "focus": false,
  "lastAction": "zoom",
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;

exports[`zoom lifecycle 2`] = `
Object {
  "SVGHeight": 200,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 200,
  "a": 4,
  "b": 0,
  "c": 0,
  "d": 4,
  "e": 0,
  "endX": null,
  "endY": null,
  "f": 0,
  "focus": false,
  "lastAction": "zoom",
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;

exports[`zoom test zoom in 1`] = `
Object {
  "SVGHeight": 400,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 400,
  "a": 2,
  "b": 0,
  "c": 0,
  "d": 2,
  "e": 0,
  "endX": null,
  "endY": null,
  "f": 0,
  "focus": false,
  "lastAction": "zoom",
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;

exports[`zoom test zoom in 2`] = `
Object {
  "SVGHeight": 400,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 400,
  "a": 4,
  "b": 0,
  "c": 0,
  "d": 4,
  "e": 0,
  "endX": null,
  "endY": null,
  "f": 0,
  "focus": false,
  "lastAction": "zoom",
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;

exports[`zoom test zoom out 1`] = `
Object {
  "SVGHeight": 400,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 400,
  "a": 0.5,
  "b": 0,
  "c": 0,
  "d": 0.5,
  "e": 0,
  "endX": null,
  "endY": null,
  "f": 0,
  "focus": false,
  "lastAction": "zoom",
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;

exports[`zoom test zoom out 2`] = `
Object {
  "SVGHeight": 400,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 400,
  "a": 0.25,
  "b": 0,
  "c": 0,
  "d": 0.25,
  "e": 0,
  "endX": null,
  "endY": null,
  "f": 0,
  "focus": false,
  "lastAction": "zoom",
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;
